### ISSUE TITLE
Investigate feasibility of implementing a layout system

### ISSUE THREAD

**Issue by kasperisager** [Siteimprove/alfa#183]
All APIs that deal with visuals would benefit from having access to a layout system in addition to the style system already implemented. While the style system can answer questions such as "does this element participate in layout?", we need a layout system to answer questions such as "where in the layout does this element end up?".

At the very least, having access to a [box tree](https://drafts.csswg.org/css-display/#box-tree) would allow us to do these things:

- Determine if an element has been positioned outside the viewport. With just the style system, we are forced to rely on heuristics for this, which is fortunately good enough for the majority of cases.

- Determine the node stack, and paint order of this stack, at a given region of the document. This would allow us to cover cases of fore-/background colour computation that are not possible based on the style system alone.

**Comment by kasperisager** [comment]
After much consideration, I have concluded that it's not feasible for us to implement a layout system. It's simply too complex and error prone if we don't get it right. Similar to DOM and CSSOM construction and JavaScript evaluation, we'll instead leave it to callers to pass in layout information if they have it available. Combined with information from the style system, we can then construct a box tree.

In summary, we'll be handling this by extending the DOM types with layout information to be filled in by callers.

**Comment by kasperisager** [comment]
Closing in favour of #756.
